### PR TITLE
feat(ui): setup authentication using google oauth2.0

### LINKED ISSUE #72: [UI] Google OAuth2 for login

### PR DESCRIPTION
**Pull request by Aksh-Bansal-dev**
closes #72

### CODE REVIEW

File: ui/src/api/googleOauth.ts
Comment (Samy-33) [inline ui/src/api/googleOauth.ts:23]: When I read this function name, I have many questions:
1. Who are we sending the google access token to?
2. Why are we sending google access token?

Let's rename this so these two questions are answered by the name:
Something like: `convertAccessToken` and have a JSDoc to explain what's happening here i.e. exchanging google's access_token for our API's access token.

File: ui/src/api/googleOauth.ts
Comment (Samy-33) [inline ui/src/api/googleOauth.ts:24]: This will be renamed to googleAccessToken, with the above mentioned changes.

File: ui/src/api/googleOauth.ts
Comment (Samy-33) [inline ui/src/api/googleOauth.ts:20]: [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) can be used to make it more readable. It will encode the parameters as well.:
```ts
const queryParams = new URLSearchParams({
  scope: 'https://www.googleapis.com/auth/userinfo.email,https://www.googleapis.com/auth/userinfo.profile'
...
});

window.location.href = `https://accounts.google.com/o/oauth2/v2/auth?${queryParams.toString()}`;
```

File: ui/src/pages/LoginRedirect/index.tsx
Comment (Samy-33) [inline ui/src/pages/LoginRedirect/index.tsx:14]: Just `!access_token` should suffice.

File: ui/src/pages/LoginRedirect/index.tsx
Comment (Samy-33) [inline ui/src/pages/LoginRedirect/index.tsx:17]: Allowing `sendGoogleAccessToken` to use `navigate` gives it too much control. I would create a helper method `redirectToRoot()` and send it instead.

```ts
const redirectToRoot = () => navigate('/');
...
sendGoogleAccessToken(access_token, redirectToRoot);
```

File: ui/src/pages/LoginRedirect/index.tsx
Comment (Samy-33) [inline ui/src/pages/LoginRedirect/index.tsx:10]: `access_token` --> `accessToken`

File: ui/src/utils/token.ts
Comment (Samy-33) [inline ui/src/utils/token.ts:10]:
```suggestion
  return !!token;
```

**Review by Samy-33 — CHANGES_REQUESTED** [review]
Please add tests for this functionality.

**Comment by Aksh-Bansal-dev** [comment]
> Please add tests for this functionality.

Do you mean adding tests for the `ProtectedRoute` component?

**Comment by Samy-33** [comment]
Test for the whole authentication feature. You'll write tests for Login Component and LoginRedirect Component.

**Comment by Aksh-Bansal-dev** [comment]
> Test for the whole authentication feature. You'll write tests for Login Component and LoginRedirect Component.

How would you write unit tests for these components? Wouldn't it be better to write an end-to-end test for this feature?

**Comment by Samy-33** [comment]
Examples of a unit test case:
- When user clicks on login, url should change to google login with correct params
- When token is already set and is valid, login page should redirect to `/`

etc..

File: ui/src/api/googleOauth.ts
Comment (Samy-33) [inline ui/src/api/googleOauth.ts:51]: Error should be handled here just in case `access_token` is invalid or the API throws error due to some other reason.

**Review by Samy-33 — CHANGES_REQUESTED** [review]
We should have unittests along with the feature in a single PR. This leaves no room for procrastination as we usually don't take them seriously. Tests shouldn't take much time. So, merge the changes from the `test` branch to this one when you're done.

**Review by Samy-33 — APPROVED** [review]
I still see some issues and tests don't cover all the usecases.

I am approving this as it's been pending for a few days now. But considering refactoring and adding more tests. Let me know if you'd like me to add some tests as examples.

**Comment by Aksh-Bansal-dev** [comment]
> Let me know if you'd like me to add some tests as examples.

That would be great.

**Comment by Aksh-Bansal-dev** [comment]
Should we merge this PR?

**Comment by Samy-33** [comment]
Go ahead and merge this, but make sure after #84 is merged, you make corresponding changes in the logic.

**Comment by Samy-33** [comment]
Also, please create an issue to add tests for Login and LoginRedirect components.

**Pull request merged**
